### PR TITLE
Added ENUM and SET types for MySQL

### PR DESCRIPTION
**Pull request by bartekrylko**
Added as saparate PR - based on #452

### CODE REVIEW

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 1]
```diff
@@ -1052,4 +1058,17 @@ public function describeTable($tableName)
...
     }
+
+    /**
+     * Returns MySQL column types (inherited and MySQL specified).
+     * @return array
+     */
+    public function getColumnTypes()
+    {
+        return array_merge(parent::getColumnTypes(),
+            [
```
Comment (shadowhand) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: This needs to be PHP 5.3 compatible, not using short array syntax.

File: src/Phinx/Db/Table/Column.php [hunk 2]
```diff
@@ -455,6 +460,34 @@ public function getProperties()
...
+     *
+     * @param mixed (array|string) $values
+     *
+     * @return Column
+     */
+    public function setValues($values)
+    {
+        if(is_array($values))
+        {
+            foreach($values as &$value) { $value = "'{$value}'"; }
```
Comment (shadowhand) [inline src/Phinx/Db/Table/Column.php]: Quoting should not be done on set, but on write.

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php [hunk 3]
```diff
@@ -576,7 +576,9 @@ public function testGetColumns()
               ->addColumn('column17', 'point')
               ->addColumn('column18', 'linestring')
               ->addColumn('column19', 'polygon')
-              ->addColumn('column20', 'uuid');
+              ->addColumn('column20', 'uuid')
+              ->addColumn('column21', 'set', array('values' => "'one', 'two'"))
+              ->addColumn('column22', 'enum', array('values' => ['three', 'four']));
```
Comment (shadowhand) [inline tests/Phinx/Db/Adapter/MysqlAdapterTest.php]: short array needs to be long array

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php [hunk 4]
```diff
@@ -621,7 +623,9 @@ public function testGetColumnsReservedTableName()
...
```
Comment (shadowhand) [inline tests/Phinx/Db/Adapter/MysqlAdapterTest.php]: short array

**Comment by bartekrylko** [comment]
Please take a look on that changes.

**Comment by alexbowers** [comment]
Any update on enum support?

**Comment by shadowhand** [comment]
Sorry I started a new job and didn't have time to go through PRs. Am reviewing now.

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 5]
```diff
@@ -960,6 +965,10 @@ protected function getColumnSqlDefinition(Column $column)
         } elseif (isset($sqlType['limit'])) {
             $def .= '(' . $sqlType['limit'] . ')';
         }
+        if(($values = $column->getValues()) && is_array($values)) {
+            foreach($values as &$value) { $value = "'{$value}'"; }
+            $def .= '('.implode(',', $values).')';
```
Comment (shadowhand) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: I would prefer these two lines be merged into one:

```php
$def .= "('" . implode("', '", $values) . "')";
```

File: src/Phinx/Db/Table/Column.php [hunk 6]
```diff
@@ -455,6 +460,30 @@ public function getProperties()
...
     /**
+     * Sets field values.
+     *
+     * @param mixed (array|string) $values
+     *
+     * @return Column
+     */
+    public function setValues($values)
+    {
+        if(!is_array($values)) { $values = explode(',', $values); }
```
Comment (shadowhand) [inline src/Phinx/Db/Table/Column.php]: This needs to match PSR-2 standards:

```php
if (...) {
    ...
}
```

**Comment by bartekrylko** [comment]
Updated. Please take a look.

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 7]
```diff
@@ -960,6 +965,9 @@ protected function getColumnSqlDefinition(Column $column)
         } elseif (isset($sqlType['limit'])) {
             $def .= '(' . $sqlType['limit'] . ')';
         }
+        if(($values = $column->getValues()) && is_array($values)) {
```
Comment (shadowhand) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: Still slightly off for PSR compatibility, but close enough. (Need `if (` with a space.)

File: src/Phinx/Db/Table/Column.php [hunk 8]
```diff
@@ -105,6 +105,11 @@ class Column
     protected $properties = array();
 
     /**
+     * @var string
```
Comment (shadowhand) [inline src/Phinx/Db/Table/Column.php]: type should be an array

File: src/Phinx/Db/Table/Column.php [hunk 9]
```diff
@@ -455,6 +460,32 @@ public function getProperties()
...
+     * Sets field values.
+     *
+     * @param mixed (array|string) $values
+     *
+     * @return Column
+     */
+    public function setValues($values)
+    {
+        if(!is_array($values)) {
+            $values = explode(',', $values);
```
Comment (shadowhand) [inline src/Phinx/Db/Table/Column.php]: I would use `preg_split('/,\s*/', $values)` to account for `foo, bar, baz`

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php [hunk 10]
```diff
@@ -621,7 +623,9 @@ public function testGetColumnsReservedTableName()
               ->addColumn('column17', 'point')
               ->addColumn('column18', 'linestring')
               ->addColumn('column19', 'polygon')
-              ->addColumn('column20', 'uuid');
+              ->addColumn('column20', 'uuid')
+              ->addColumn('column21', 'set', array('values' => "one, two"))
```
Comment (shadowhand) [inline tests/Phinx/Db/Adapter/MysqlAdapterTest.php]: this example will be slightly broken because of the issue with spaces in a comma separated list

**Comment by shadowhand** [comment]
I think the comma-separated list problem should be fixed before merge. Otherwise everything looks good.

**Comment by bartekrylko** [comment]
I think, current approach is not a bug. MySQL allows for ENUM/SET values with white space like ' one'. Why should we "fix" such cases?

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 11]
```diff
@@ -960,6 +965,9 @@ protected function getColumnSqlDefinition(Column $column)
...
+            $def .= "('" . implode("', '", $values) . "')";
```
Comment (alexbowers) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: As per @shadowhand comment below about using preg_split

**Comment by shadowhand** [comment]
> MySQL allows for ENUM/SET values with white space like ' one'. 

That's absolutely true, but I think the expectation for a string list of values `"a, b, c"` would be that it produce `"a", "b", "c"` and not `"a", " b", " c"`.

**Comment by bartekrylko** [comment]
I'm still not 100% sure about that change in Column class. But updated as you suggested. Hopefully, expectations are as you mentioned.

**Comment by shadowhand** [comment]
Thanks!

**Comment by bartekrylko** [comment]
thanks

**Comment by alexbowers** [comment]
Any chance of a documentation PR @bartekrylko

**Comment by bartekrylko** [comment]
Sure, here you go. Two simple examples:
```
->addColumn('enum_col_name', 'enum', array('values' => "one, two", 'default' => 'one'))
->addColumn('set_col_name', 'set', array('values' => array('one','two'), 'null' => true))
```
For both types (ENUM and SET), 'values' option supports string and array.

**Comment by alexbowers** [comment]
The string being comma separated with or without spaces?

Is this feature MySQL only. Or supported in all drivers?

**Comment by alexbowers** [comment]
#474 Has provided proper documentation for it.

**Comment by bartekrylko** [comment]
This is supported for MySQL only.
> The string being comma separated with or without spaces? @alexbowers 

Doesn't matter. Value will be fetched without spaces anyway.

**Pull request merged**